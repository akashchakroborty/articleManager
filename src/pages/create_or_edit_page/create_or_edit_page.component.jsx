import React from "react";
import { CreateOrEditPageContainer, Back } from "./create_or_edit_page.styles";
import CreateArticle from "../../components/create_article/create_article.component";
import EditArticle from "../../components/edit_article/edit_article.component";
import WithSpinner from "../../components/with-spinner/with-spinner.component";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import {
  selectIsFetching,
  selectArticle,
  selectArticles,
} from "../../redux/articles/articles.selectors";
import NotFound from "../../components/notfound/notfound.component";

export const CreateOrEditPage = ({
  isLoading,
  articles,
  article,
  match,
  history,
}) => {
  const validateArticleEditRoute = () => {
    const validIds = articles.map((val) => val.id);
    return validIds.includes(match.params.id) || !article;
  };

  return validateArticleEditRoute() ? (
    <CreateOrEditPageContainer>
      {article ? <EditArticle article={article} /> : <CreateArticle />}
      <Back className="back-button" onClick={() => history.push("/")}>
        BACK
      </Back>
    </CreateOrEditPageContainer>
  ) : (
    <NotFound history={history} />
  );
};

const mapStateToProps = createStructuredSelector({
  isLoading: selectIsFetching,
  article: selectArticle,
  articles: selectArticles,
});

export default connect(mapStateToProps)(WithSpinner(CreateOrEditPage));
