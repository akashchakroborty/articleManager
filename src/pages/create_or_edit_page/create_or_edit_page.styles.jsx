import styled from "styled-components";

export const CreateOrEditPageContainer = styled.div`
  width: 850px;
  display: flex;
  justify-content: center;
  margin: 30px auto;
  @media screen and (max-width: 800px) {
    flex-direction: column;
    width: unset;
    align-items: center;
    > *:first-child {
      margin-bottom: 50px;
    }
  }
`;

export const Back = styled.div`
  text-decoration: none;
  border: 1px solid grey;
  cursor: pointer;
  border-radius: 4px;
  padding: 4px 12px;
  width: fit-content;
  align-self: flex-end;
  color: grey;
  letter-spacing: 3px;
  font-weight: bold;
  line-height: 30px;
  @media screen and (max-width: 800px) {
    margin-right: 20px;
  }
`;

Back.displayName = "Back";
