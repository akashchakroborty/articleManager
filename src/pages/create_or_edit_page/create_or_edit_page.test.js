import React from "react";
import { shallow } from "enzyme";
import { CreateOrEditPage } from "./create_or_edit_page.component";

describe("CreateOrEditPage component", () => {
  let wrapper;
  let mockHistory;
  let mockMatch;

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };
    mockMatch = {
      params: {
        id: "1234",
      },
    };
    const mockProps = {
      articles: [
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "1234",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
      ],
      article: {
        title: "Mock data",
        description: "Mock data",
        author: "Mock data",
        content: "Mock data",
        id: "1234",
        img: "Mock data",
        avatar: "Mock data",
        claps: "Mock data",
      },
      history: mockHistory,
      match: mockMatch,
    };

    wrapper = shallow(<CreateOrEditPage {...mockProps} />);
  });

  it("should render CreateOrEditPage component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call history.push when Back is clicked", () => {
    wrapper.find("Back").simulate("click");
    expect(mockHistory.push).toHaveBeenCalled();
  });

  it("should render Not Found when we have an invalid route.", () => {
    const newMockProps = {
      articles: [
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "12",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
      ],
      article: {
        title: "Mock data",
        description: "Mock data",
        author: "Mock data",
        content: "Mock data",
        id: "12",
        img: "Mock data",
        avatar: "Mock data",
        claps: "Mock data",
      },
      history: mockHistory,
      match: mockMatch,
    };
    const newWrapper = shallow(<CreateOrEditPage {...newMockProps} />);

    expect(newWrapper.exists("NotFound")).toBe(true);
  });

  it("should render Create Article when article is null.", () => {
    const newMockProps = {
      articles: [
        {
          title: "Mock data",
          description: "Mock data",
          author: "Mock data",
          content: "Mock data",
          id: "12",
          img: "Mock data",
          avatar: "Mock data",
          claps: "Mock data",
        },
      ],
      article: null,
      history: mockHistory,
      match: mockMatch,
    };
    const newWrapper = shallow(<CreateOrEditPage {...newMockProps} />);
    expect(newWrapper.exists("withRouter(Connect(CreateArticle))")).toBe(true);
  });
});
