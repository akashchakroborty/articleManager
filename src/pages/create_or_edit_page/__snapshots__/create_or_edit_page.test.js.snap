// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOrEditPage component should render CreateOrEditPage component 1`] = `ShallowWrapper {}`;
